lib/Storage/S3StorageClient.cpp

```diff
@@ -0,0 +1,379 @@
+                                   Source.str().c_str());
+
+  FilenameMap[Destination] = FilenameMap[Source];
+  return llvm::Error::success();
+}
+
+llvm::Expected<ReadableFile>
+S3StorageClient::getReadableFile(const llvm::StringRef Path) {
+  using llvm::sys::fs::createTemporaryFile;
+  if (FilenameMap.count(Path) == 0)
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "File %s does not exist",
+                                   Path.str().c_str());
+
+  Aws::S3::Model::GetObjectRequest Request;
+  Request.SetBucket(Bucket);
+  Request.SetKey(resolvePath(SubPath, FilenameMap[Path]));
+
+  Aws::S3::Model::GetObjectOutcome Result = Client.GetObject(Request);
+  if (not Result.IsSuccess())
+    return toError(Result);
+
+  llvm::SmallString<128> TempFile;
+  std::error_code EC = createTemporaryFile("revng-s3-storage-", "", TempFile);
+  if (EC)
+    return llvm::createStringError(EC, "Could not create temporary file");
+
+  std::ofstream OS(TempFile.str().str(), std::ios::binary);
+  if (OS.fail())
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "Could not open temporary file");
+
+  constexpr std::size_t BufSize = 4096;
+  char Buffer[BufSize];
+  auto &Body = Result.GetResult().GetBody();
+
+  while (!Body.eof()) {
+    Body.read(Buffer, BufSize);
+    OS.write(Buffer, Body.gcount());
+  }
+
+  OS.flush();
+  OS.close();
+
+  auto MaybeReadableStream = llvm::MemoryBuffer::getFile(TempFile);
+  if (not MaybeReadableStream)
+    return llvm::createStringError(MaybeReadableStream.getError(),
+                                   "Failed to open the file for reading");
+
+  return ReadableFile{ std::move(MaybeReadableStream.get()),
+                       [TempFile]() { llvm::sys::fs::remove(TempFile); } };
+}
+
+llvm::Expected<WritableFile>
+S3StorageClient::getWritableFile(const llvm::StringRef Path,
+                                 ContentEncoding Encoding) {
+  using llvm::sys::fs::createTemporaryFile;
+
+  std::string PathCopy = Path.str();
+  llvm::SmallString<128> TempFile;
+  int TempFileFD;
+  std::error_code EC = createTemporaryFile("revng-s3-storage-",
+                                           "",
+                                           TempFileFD,
+                                           TempFile);
+  if (EC)
+    return llvm::createStringError(EC, "Could not create temporary file");
+
+  auto OS = std::make_unique<llvm::raw_fd_ostream>(TempFileFD, true);
+  return WritableFile{
+    std::move(OS),
+    [this, TempFile, PathCopy, Encoding]() -> llvm::Error {
+      Aws::S3::Model::PutObjectRequest Request;
+      Request.SetBucket(Bucket);
+      std::string NewFilename = generateNewFilename(PathCopy);
+      Request.SetKey(resolvePath(SubPath, NewFilename));
+
+      if (Encoding == ContentEncoding::Gzip)
+        Request.SetContentEncoding("gzip");
+
+      auto File = Aws::MakeShared<Aws::FStream>("ALLOC_TAG",
```

Comment:
What is `ALLOC_TAG`?

Reply:
Re-checking again, it seems you need to use `Aws::MakeShared` only if you use a custom allocator (which we don't)
I've replaced all occurrencies with `std::make_shared`